text/3874-build-std-always.md

```diff
@@ -0,0 +1,1543 @@
+proposed changes.
+
+Cargo configuration will contain a new key `build-std` under the `[build]`
+section ([?][rationale-build-std-in-config]), permitting one of two values -
+"never" ([?][rationale-build-std-never]) or "always", defaulting to "never":
+
+```toml
+[build]
+build-std = "never" # or `always`
+```
+
+`build-std` can also be specified in the `[target.<triple>]` and
+`[target.<cfg>]` sections ([?][rationale-build-std-target-section]):
+
+```toml
+[target.aarch64-unknown-illumos]
+build-std = "never" # or `always`
+```
+
+The `build-std` configuration locations have the following precedence
+([?][rationale-build-std-precedence]):
+
+1. `[target.<triple>]`
+2. `[target.<cfg>]`
+3. `[build]`
+
+As the Cargo configuration is local to the current user (typically in
+`.config/cargo.toml` in the project root and/or Cargo home directory), the value
+of `build-std` is not influenced by the dependencies of the current crate.
+
+When `build-std` is set to "always", then the standard library will be
+unconditionally recompiled ([?][rationale-unconditional]) in the release profile
+defined in its workspace as part of every clean build
+([?][rationale-release-profile]). This is primarily useful for users of tier
+three targets. As with other dependencies, the standard library's build will
+respect the `RUSTFLAGS` environment variable.
+
+> [!NOTE]
+>
+> Configuration of the pre-built standard library is split across bootstrap and
+> the Cargo packages for the standard library. As much of this configuration as
+> possible should be moved to the Cargo profile for these packages so that the
+> artifacts produced by build-std match the pre-built standard library as much
+> as is feasible.
+
+`build-std` is a short-hand for an object which sets the `when` key:
+
+```toml
+[build]
+build-std = { when = "always" }
+```
+
+Alongside `build-std`, a `build-std.crates` key will be introduced
+([?][rationale-build-std-crate]), which can be used to specify which crates from
+the standard library should be built. Only "core", "alloc" and "std" are valid
+values for `build-std.crates`.
+
+```toml
+[build]
+build-std = { when = "always", crates = "std" }
+```
+
+A value of "std" means that every crate in the graph has a direct dependency on
+`std`, `alloc` and `core`. Similarly, "alloc" means `alloc` and `core`, and
+"core" means just `core`.
+
+If `std` is to be built and Cargo is building a test or benchmark using the
+default test harness then Cargo will also build the `test` crate.
+
+If [*Standard library dependencies*][rfcs#3875] are implemented then `builtin`
+dependencies will be used if `build-std.crates` is not explicitly set.
+Otherwise, `build-std.crates` will default to the crate intended to be supported
+by the target (see later
+[*Standard library crate stability*][standard-library-crate-stability] section).
+
+> [!NOTE]
+>
+> Inspired by the concept of [opaque dependencies][Opaque dependencies], the
```

Comment:
@nicoburns brought up on reddit the idea of `cargo clean` not affecting opaque dependencies, requiring some other step to clean them (maybe a `cargo clean -p std`).  Alternatively, we may get this "for free" with rust-lang/cargo#5931.

I lean towards keeping `cargo clean` the same for now but might be worth calling this question out somewhere.